rfcs/20220128-f64-data_type/Readme.md

```diff
@@ -0,0 +1,72 @@
+this requirement by introducing 64-bit, double precision float convolution support in oneDNN.
+We propose adding f64 data type support for convolution on the GPU. Since `reorder` is used to support convolution with different tensor formats, fp64 support will be added to this primitive as well. Other primitives, or engines are not impacted at this point; although library infrastrcture is modified to recognize double-precision floating-point data type.
+
+## API changes
+
+Here's the list of of the changes in library header files:
+
+include/oneapi/dnnl/dnnl.hpp:
+
+```c
+    /// Data type specification.
+    enum class data_type {
+        /// Undefined data type (used for empty memory descriptors).
+        undef = dnnl_data_type_undef,
+        /// [16-bit/half-precision floating point](https://en.wikipedia.org/wiki/Half-precision_floating-point_format).
+        f16 = dnnl_f16,
+        /// non-standard
+        /// [16-bit floating point with 7-bit mantissa](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format).
+        bf16 = dnnl_bf16,
+        /// [32-bit/single-precision floating point](https://en.wikipedia.org/wiki/Single-precision_floating-point_format).
+        f32 = dnnl_f32,
+        //// [64-bit/double-precision floating point](https://en.wikipedia.org/wiki/Double-precision_floating-point_format).
+        f64 = dnnl_f64,
+        /// 32-bit signed integer.
+        s32 = dnnl_s32,
+        /// 8-bit signed integer.
+        s8 = dnnl_s8,
+        /// 8-bit unsigned integer.
+        u8 = dnnl_u8,
+    };
+```
+
+include/oneapi/dnnl/dnnl_types.h:
+
+```c
+    typedef enum {
+        /// Undefined data type, used for empty memory descriptors.
+        dnnl_data_type_undef = 0,
+        /// 16-bit/half-precision floating point.
+        dnnl_f16 = 1,
+        /// non-standard 16-bit (bfloat16 w/ 7 bit mantissa) floating point.
+        dnnl_bf16 = 2,
+        /// 32-bit/single-precision floating point.
+        dnnl_f32 = 3,
+        /// 32-bit signed integer.
+        dnnl_s32 = 4,
+        /// 8-bit signed integer.
+        dnnl_s8 = 5,
+        /// 8-bit unsigned integer.
+        dnnl_u8 = 6,
+        /// 64-bit/double-precision floating point.
+        dnnl_f64 = 7,
+    } dnnl_data_type_t;
+```
+
+## Some details on Convolution primitive
+
+Convolution primitive in particular has several configurations and parameters that deserve to be discussed in more detail with regards to f64 data type.
+- Table in https://oneapi-src.github.io/oneDNN/dev_guide_convolution.html describes the supported configurations in the convolution api. We propose adding one row, for which source, weights, destination and bias are all f64; for both forward and backward propagation.
+- Data Representation: f64 conv will be supported for all the representations which f32 is supported on.
+- Postops & attributes: all post-ops and attributes supported for f32 configuration will be supported for f64 on GPU.
```

Comment:
I think it's worth mentioning that all post-op operations in `f64` case are to be applied in `f64` precision. Now, for any data types, post-ops are always done in `f32`: https://oneapi-src.github.io/oneDNN/dev_guide_data_types.html#general-numerical-behavior-of-the-onednn-library

Reply:
@echeresh @hkianine Will it be enough to have post-op parameters like alpha and beta in f32 data type?

Reply:
Yes. scale parameters like alpha and beta are passed as f32 (so we don't have to change APIs), and inside the kernel they'll be upconverted to f64 and math will be done in f64.

Reply:
I think we don't have specific requests/cases to have post-op parameters in `f64`. So it should be fine to leave them as is. If requested later, we can introduce new interfaces with `double` in the future.

Reply:
If there is no request for f64 post-ops, I would advocate to leave them out of the scope of the work for now (so no postops for f64). This is for 2 reasons:
- post ops are mostly useful for inference, and f64 will likely be mostly used for training here. They don't even know if f64 is actually required for training per their own words (they did not try to use f32 for the DNN portion of their workload), so here f64 more looks like a commodity than anything else.
- we don't have a definition of which accuracy is needed for elementary functions (sigmoid, tanh, ...), and accurate double precision eltwise can be costly to compute/implement.

Reply:
I agree. I added a note that for f64 config, math will occur in f64, even though user passes parameters (such as scales) in f32. 
We can revisit this topic once we collect user's feedback.

Reply:
+1 to Mourad's suggestion. Looks like initial f64 support is going to be limited anyway (no CPU support, only convolution/reorder) - then why not to limit it further and skip post-ops. This makes sense to me as we see post-ops support being an extra burden without specific requests.

Reply:
I had not paid a lot of attention to post-ops so far; mostly because some of the tests I had enabled had post-ops and they passed; of course reference cpu was in f32. Therefore I did not dig further to find out the true precision for post-ops calculation. 
Anyways, I agree with your suggestions and I updated the doc to mention no post-ops support provided for f64 config at this point. Thanks.